V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-CODE/MASWE-0046.md

Reviewer: DONE

---
title: Use of Deprecated APIs or Functionality
id: MASWE-0046
alias: deprecated-apis
requirement: "The app does not use deprecated APIs or functionality."
platform: [android, ios]
profiles: [L2]
threat: MAS-THREAT-0046
attacks: [MAS-ATTACK-0058]
mappings:
  masvs-v1: [MSTG-CRYPTO-4]
  masvs-v2: [MASVS-CODE-3, MASVS-CRYPTO-2]
  cwe: [327, 477, 522]
  maswe-beta: [MASWE-0015]
refs:
- https://developer.android.com/about/versions/12/behavior-changes-all#bouncy-castle
- https://developer.android.com/reference/java/security/KeyStore
- https://labs.withsecure.com/publications/how-secure-is-your-android-keystore-authentication
status: new
---

## Overview

This weakness occurs when an app relies on deprecated APIs or functionality that are no longer maintained, may lack security fixes, or have been superseded by safer alternatives.

> Platforms deprecate APIs precisely because better-designed replacements exist, often fixing security shortcomings of the old API. Continuing to use deprecated security-relevant APIs can leave the app exposed to weaknesses the platform has already addressed. A representative example is the use of deprecated Android KeyStore implementations such as Bouncy Castle (BKS), but this weakness covers deprecated APIs and functionality in general, including cryptography, storage, networking, and platform features.

Reviewer: ```suggestion
Platforms deprecate APIs for various reasons. It may be architectural changes, lack of developer adoption, reduced hardware support, or a better-designed replacement that fixes security shortcomings of the old API. Continuing to use deprecated security-relevant APIs can leave the app exposed to weaknesses the platform has already addressed. A representative example is the use of deprecated Android KeyStore implementations such as Bouncy Castle (BKS), but this weakness covers deprecated APIs and functionality in general, including cryptography, storage, networking, and platform features.
```

REMARK: I would not use the word "precisely" because it sounds like we are referencing something not mentioned before.

## Modes of Introduction

- **Deprecated Cryptographic Providers**: Using deprecated keystore or cryptographic provider implementations (e.g. Bouncy Castle-backed KeyStore on Android) instead of their maintained replacements.
- **Deprecated Platform APIs**: Continuing to use platform APIs that have been superseded by safer alternatives with improved security behavior.
- **Deprecation Warnings Ignored**: Building with unresolved deprecation warnings, so the use of outdated functionality accumulates unnoticed.

> ## Impact

Reviewer: Theoretically, deprecated APIs can be fully removed at any time. Should the app rely on this API (and not just the security-relevant deprecated one), it could cause runtime errors and even crashes when they are not handled properly.

This would be a denial of service, which could result in financial damage to the developer. 

I assume this is pretty rare, but should we mention it here?

Author: We won't consider this for now, it's more functional testing than anything else.

- **Compromise of Sensitive Data**: Attackers can exploit the known shortcomings of deprecated implementations, resulting in exposure of data those APIs were supposed to protect.
- **Authentication or Authorization Bypass**: Attackers can abuse weaknesses in deprecated authentication- or keystore-related APIs, resulting in unauthorized access to protected keys or functionality.

## Mitigations

> - **Migrate Off Deprecated APIs**: Replace deprecated security-relevant APIs with the platform's recommended alternatives as part of regular maintenance.

Reviewer: ```suggestion
- **Migrate Off Deprecated APIs**: Replace deprecated APIs with the platform's recommended alternatives as part of regular maintenance. Prioritize migration of security-relevant APIs.
```

- **Treat Deprecation Warnings as Actionable**: Surface deprecation warnings in builds and reviews, and track migrations for security-relevant APIs with priority.

> - **Follow Platform Release Notes**: Review each platform release's deprecation and behavior-change notes to plan migrations before old functionality is removed or breaks.

Reviewer: ```suggestion
- **Follow Platform Release Notes**: Review each platform release's deprecation and behavior-change notes to plan migrations before old functionality is removed or breaks.
- **Test the App on All Supported Platforms**: Run the app on all platform API versions supported during development.
```
